### ISSUE TITLE
JS slices-in-struct-fields may be broken

### ISSUE THREAD

**Issue by Manishearth** [rust-diplomat/diplomat#634]
When there are slices in a struct field, `_intoFFI` contains this:

https://github.com/rust-diplomat/diplomat/blob/9320edea0fe21bfcdcf19918724fd6bbf88b541d/feature_tests/js/api/BorrowedFields.mjs#L48

These are JS objects, not FFI compatible values. We should probably add a `.splat()` function to DiplomatBuf that returns `[ptr, size]`.

Unfortunately, due to other bugs in the struct impl, I have been unable to test this.

**Comment by Manishearth** [comment]
cc @ambiguousname, though I plan on fixing this (I just can't _test_ it easily right now)

**Comment by ambiguousname** [comment]
Alright! This is definitely my fault, I assumed whatever tests I created in slices.mjs would catch these sorts of issues while I was adding these helper classes. I can make a fix tomorrow morning if the issue is still open then,

**Comment by Manishearth** [comment]
@ambiguousname Also I think the slice cleanup callbacks here do the wrong thing. They should be cleaning up these strs8 (etc) values, instead they clean up the fields (which will not have free/garbagecollect methods)

https://github.com/rust-diplomat/diplomat/blob/9320edea0fe21bfcdcf19918724fd6bbf88b541d/feature_tests/js/api/BorrowedFields.mjs#L42-L48

I've reassigned to you since this part is a bit more complex; I could still try and figure it out but I'm not going to go that deep tonight. If you end up not having time to, I can still look at it.

I think the best game plan here may to allow the conversion code to pre-append slice conversions, similar to what we do in methods. Ideally this code can be shared between structs and methods but it might not be possible.

**Comment by Manishearth** [comment]
The fix for the first issue is not as simple as adding a `.splat()`, because we call `gen_js_to_c_for_type` in both methods (where the code expects it to be a DiplomatBuf) and in struct fields (where the code wants to pass it directly to FFI). Probably having the code append a conversion to some passed-in array that then gets used by both methods and fields is sufficient, then the impl of `gen_js_to_c_for_type` can return `convertedSlice.ptr, convertedSlice.size`

Patch included for posterity:

<details>

```patch
commit 9788c9613c9abc9e0d82686a866a0207376cf98c
Author: Manish Goregaokar <manishsmail@gmail.com>
Date:   Thu Aug 15 19:07:38 2024 -0700

    Correctly splat JS arrays in struct toFFI()s

diff --git a/tool/src/js/type_generation/converter.rs b/tool/src/js/type_generation/converter.rs
index 88c9c139..9dc07fc8 100644
--- a/tool/src/js/type_generation/converter.rs
+++ b/tool/src/js/type_generation/converter.rs
@@ -529,12 +529,14 @@ impl<'jsctx, 'tcx> TyGenContext<'jsctx, 'tcx> {
             Type::Struct(..) => self.gen_js_to_c_for_struct_type(js_name, struct_borrow_info),
             Type::Slice(hir::Slice::Str(_, encoding)) => match encoding {
                 hir::StringEncoding::UnvalidatedUtf8 | hir::StringEncoding::Utf8 => {
-                    format!("diplomatRuntime.DiplomatBuf.str8(wasm, {js_name})").into()
+                    format!("...diplomatRuntime.DiplomatBuf.str8(wasm, {js_name}).splat()").into()
+                }
+                _ => {
+                    format!("...diplomatRuntime.DiplomatBuf.str16(wasm, {js_name}).splat()").into()
                 }
-                _ => format!("diplomatRuntime.DiplomatBuf.str16(wasm, {js_name})").into(),
             },
             Type::Slice(hir::Slice::Strs(encoding)) => format!(
-                r#"diplomatRuntime.DiplomatBuf.strs(wasm, {js_name}, "{}")"#,
+                r#"...diplomatRuntime.DiplomatBuf.strs(wasm, {js_name}, "{}").splat()"#,
                 match encoding {
                     hir::StringEncoding::UnvalidatedUtf16 => "string16",
                     _ => "string8",
@@ -542,7 +544,7 @@ impl<'jsctx, 'tcx> TyGenContext<'jsctx, 'tcx> {
             )
             .into(),
             Type::Slice(hir::Slice::Primitive(_, p)) => format!(
-                r#"diplomatRuntime.DiplomatBuf.slice(wasm, {js_name}, "{}")"#,
+                r#"...diplomatRuntime.DiplomatBuf.slice(wasm, {js_name}, "{}").splat()"#,
                 self.formatter.fmt_primitive_list_view(p)
             )
             .into(),
diff --git a/tool/templates/js/runtime.mjs b/tool/templates/js/runtime.mjs
index ff00de2f..803626f3 100644
--- a/tool/templates/js/runtime.mjs
+++ b/tool/templates/js/runtime.mjs
@@ -66,66 +66,66 @@ export function enumDiscriminant(wasm, ptr) {
  */
 export class DiplomatBuf {
     static str8 = (wasm, string) => {
-    var utf8Length = 0;
-    for (const codepointString of string) {
-        let codepoint = codepointString.codePointAt(0);
-        if (codepoint < 0x80) {
-        utf8Length += 1
-        } else if (codepoint < 0x800) {
-        utf8Length += 2
-        } else if (codepoint < 0x10000) {
-        utf8Length += 3
-        } else {
-        utf8Length += 4
+        var utf8Length = 0;
+        for (const codepointString of string) {
+            let codepoint = codepointString.codePointAt(0);
+            if (codepoint < 0x80) {
+            utf8Length += 1
+            } else if (codepoint < 0x800) {
+            utf8Length += 2
+            } else if (codepoint < 0x10000) {
+            utf8Length += 3
+            } else {
+            utf8Length += 4
+            }
         }
-    }
 
-    const ptr = wasm.diplomat_alloc(utf8Length, 1);
+        const ptr = wasm.diplomat_alloc(utf8Length, 1);
 
-    const result = (new TextEncoder()).encodeInto(string, new Uint8Array(wasm.memory.buffer, ptr, utf8Length));
-    console.assert(string.length === result.read && utf8Length === result.written, "UTF-8 write error");
+        const result = (new TextEncoder()).encodeInto(string, new Uint8Array(wasm.memory.buffer, ptr, utf8Length));
+        console.assert(string.length === result.read && utf8Length === result.written, "UTF-8 write error");
 
-    return new DiplomatBuf(ptr, utf8Length, () => wasm.diplomat_free(ptr, utf8Length, 1));
+        return new DiplomatBuf(ptr, utf8Length, () => wasm.diplomat_free(ptr, utf8Length, 1));
     }
 
     static str16 = (wasm, string) => {
-    const byteLength = string.length * 2;
-    const ptr = wasm.diplomat_alloc(byteLength, 2);
+        const byteLength = string.length * 2;
+        const ptr = wasm.diplomat_alloc(byteLength, 2);
 
-    const destination = new Uint16Array(wasm.memory.buffer, ptr, string.length);
-    for (let i = 0; i < string.length; i++) {
-        destination[i] = string.charCodeAt(i);
-    }
+        const destination = new Uint16Array(wasm.memory.buffer, ptr, string.length);
+        for (let i = 0; i < string.length; i++) {
+            destination[i] = string.charCodeAt(i);
+        }
 
-    return new DiplomatBuf(ptr, string.length, () => wasm.diplomat_free(ptr, byteLength, 2));
+        return new DiplomatBuf(ptr, string.length, () => wasm.diplomat_free(ptr, byteLength, 2));
     }
 
     static slice = (wasm, list, rustType) => {
-    const elementSize = rustType === "u8" || rustType === "i8" || rustType === "boolean" ? 1 :
-        rustType === "u16" || rustType === "i16" ? 2 :
-        rustType === "u64" || rustType === "i64" || rustType === "f64" ? 8 :
-            4;
-
-    const byteLength = list.length * elementSize;
-    const ptr = wasm.diplomat_alloc(byteLength, elementSize);
-
-    /** 
-     * Create an array view of the buffer. This gives us the `set` method which correctly handles untyped values
-     */
-    const destination =
-        rustType === "u8" || rustType === "boolean" ? new Uint8Array(wasm.memory.buffer, ptr, byteLength) :
-        rustType === "i8" ? new Int8Array(wasm.memory.buffer, ptr, byteLength) :
-            rustType === "u16" ? new Uint16Array(wasm.memory.buffer, ptr, byteLength) :
-            rustType === "i16" ? new Int16Array(wasm.memory.buffer, ptr, byteLength) :
-                rustType === "i32" ? new Int32Array(wasm.memory.buffer, ptr, byteLength) :
-                rustType === "u64" ? new BigUint64Array(wasm.memory.buffer, ptr, byteLength) :
-                    rustType === "i64" ? new BigInt64Array(wasm.memory.buffer, ptr, byteLength) :
-                    rustType === "f32" ? new Float32Array(wasm.memory.buffer, ptr, byteLength) :
-                        rustType === "f64" ? new Float64Array(wasm.memory.buffer, ptr, byteLength) :
-                        new Uint32Array(wasm.memory.buffer, ptr, byteLength);
-    destination.set(list);
-
-    return new DiplomatBuf(ptr, list.length, () => wasm.diplomat_free(ptr, byteLength, elementSize));
+        const elementSize = rustType === "u8" || rustType === "i8" || rustType === "boolean" ? 1 :
+            rustType === "u16" || rustType === "i16" ? 2 :
+            rustType === "u64" || rustType === "i64" || rustType === "f64" ? 8 :
+                4;
+
+        const byteLength = list.length * elementSize;
+        const ptr = wasm.diplomat_alloc(byteLength, elementSize);
+
+        /** 
+         * Create an array view of the buffer. This gives us the `set` method which correctly handles untyped values
+         */
+        const destination =
+            rustType === "u8" || rustType === "boolean" ? new Uint8Array(wasm.memory.buffer, ptr, byteLength) :
+            rustType === "i8" ? new Int8Array(wasm.memory.buffer, ptr, byteLength) :
+                rustType === "u16" ? new Uint16Array(wasm.memory.buffer, ptr, byteLength) :
+                rustType === "i16" ? new Int16Array(wasm.memory.buffer, ptr, byteLength) :
+                    rustType === "i32" ? new Int32Array(wasm.memory.buffer, ptr, byteLength) :
+                    rustType === "u64" ? new BigUint64Array(wasm.memory.buffer, ptr, byteLength) :
+                        rustType === "i64" ? new BigInt64Array(wasm.memory.buffer, ptr, byteLength) :
+                        rustType === "f32" ? new Float32Array(wasm.memory.buffer, ptr, byteLength) :
+                            rustType === "f64" ? new Float64Array(wasm.memory.buffer, ptr, byteLength) :
+                            new Uint32Array(wasm.memory.buffer, ptr, byteLength);
+        destination.set(list);
+
+        return new DiplomatBuf(ptr, list.length, () => wasm.diplomat_free(ptr, byteLength, elementSize));
     }
 
     
@@ -155,6 +155,14 @@ export class DiplomatBuf {
         });
     }
 
+    /**
+     * Return an array with the fields suitable for passing down to a wasm function.
+     * Should typically be invoked with the spread operator
+     */
+    splat() {
+        return [this.ptr, this.size];
+    }
+
     /**
      * Generated code calls one of methods these for each allocation, to either
      * free directly after the FFI call, to leak (to create a &'static), or to
```

</details>

**Comment by ambiguousname** [comment]
For `slice_cleanup_callbacks`, this is something that I get rid of/should be fixed in #622 with the addition of slice arenas. I'll double check though.

As for the other issues, I'll take a look.

**Comment by ambiguousname** [comment]
Because of the other slice fixes in #622, I've also put a patch for these issues in there as well. Here's the commit in particular, if we want to section it off at some point: https://github.com/rust-diplomat/diplomat/pull/622/commits/7c3db89531ac5195879f111536d4f3a3b1f94a6f

**Comment by Manishearth** [comment]
Lemme prioritize landing that
